src/rust-crypto/CrossSigningInfoImpl.ts

```diff
@@ -0,0 +1,98 @@
+/*
+Copyright 2023 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { CrossSigningInfo } from "../crypto-api/CrossSigningInfo";
+import { CrossSigningKey } from "../crypto-api";
+import { IHttpOpts, MatrixHttpApi, Method } from "../http-api";
+import { IDownloadKeyResult, Keys } from "../client";
+
+/**
+ * Implementation of {@link ICrossSigningInfo}
+ */
+export class CrossSigningInfoImpl implements CrossSigningInfo {
+    /**
+     * Create an instance of {@link CrossSigningInfoImpl} for the given user
+     *
+     * @param userId - the owner of the cross signing info
+     * @param http - http interface
+     */
+    public static async create(
+        userId: string,
+        http: MatrixHttpApi<IHttpOpts & { onlyData: true }>,
+    ): Promise<CrossSigningInfoImpl | null> {
+        const queryBody = {
+            device_keys: {
+                [userId]: [],
+            },
+        };
+
+        const keyResult = await http.authedRequest<IDownloadKeyResult>(
+            Method.Post,
+            "/_matrix/client/v3/keys/query",
+            undefined,
+            queryBody,
+            {
+                prefix: "",
+            },
+        );
+
+        const getPubKey = (keys: Keys): string => {
+            // `keys` is an object with { [`ed25519:${pubKey}`]: pubKey }
+            // We assume only a single key, and we want the bare form without type
+            // prefix, so we select the values.
+            return Object.values(keys.keys)[0];
+        };
+
+        if (!keyResult.master_keys?.[userId]) return null;
+        const masterKey = getPubKey(keyResult.master_keys[userId]);
```

Comment:
any of `master_keys`, `self_signing_keys` or `user_signing_keys` could be empty, so `master_keys[userId]` may be undefined. This will throw, I think.